src/rpp/rpp/operators/combine_latest.hpp

```diff
@@ -0,0 +1,166 @@
+#include <rpp/utils/functors.hpp>
+
+#include <rpp/operators/details/subscriber_with_state.hpp> // create_subscriber_with_state
+
+
+IMPLEMENTATION_FILE(combine_latest_tag);
+
+namespace rpp::details
+{
+
+/**
+ * \brief "combine_latest" operator (an OperatorFn used by "lift").
+ */
+template<constraint::decayed_type Type, typename TCombiner, constraint::observable ...TOtherObservable>
+struct combine_latest_impl
+{
+    TCombiner m_combiner;
+    std::tuple<TOtherObservable...> m_other_observables;
+
+private:
+    static constexpr size_t INDEX_OF_SOURCE_TYPE = 0;
+
+    using DownstreamType = utils::decayed_invoke_result_t<TCombiner, Type, utils::extract_observable_type_t<TOtherObservable>...>;
+
+    /**
+     * \brief The coordinator that caches the values from all the observables and dispatches latest caches to the observer. Note the emission is only sent to the observer when all the observables at least emits once.
+     */
+    template<constraint::decayed_type SourceValueType, constraint::decayed_type ...OtherValueTypes>
+    struct combine_latest_coordinator_t
+    {
+        explicit combine_latest_coordinator_t(TCombiner combiner) : m_combiner(combiner) {}
+
+        // Not copyable nor movable.
+        combine_latest_coordinator_t(const combine_latest_coordinator_t&) = delete;
+        combine_latest_coordinator_t(combine_latest_coordinator_t&&) = delete;
+
+        template<size_t I>
+        void offer_next_value(auto&& value, const auto& subscriber)
+        {
+            if (!subscriber.is_subscribed()) return;
+
+            std::scoped_lock<std::mutex> lock{m_mutex};
+            std::get<I>(m_values) = std::forward<decltype(value)>(value);
+
+            // If all the optional(s) have value, dispatch to downstream
+            if (has_values_for_observer())
+            {
+                subscriber.on_next(combine_cached_values());
+            }
+        }
+
+        void offer_completed(const auto& subscriber)
+        {
+            if (!subscriber.is_subscribed()) return;
+
+            std::scoped_lock<std::mutex> lock{m_mutex};
+
+            if (++m_completed_count == m_total_completed_count)
+            {
+                subscriber.on_completed();
+            }
+        }
+
+    private:
+        TCombiner m_combiner;
+
+        std::mutex m_mutex;
+        std::tuple<std::optional<SourceValueType>, std::optional<OtherValueTypes>...> m_values{};
+        size_t m_completed_count{0};
+
+        const std::size_t m_total_completed_count = 1 + sizeof...(OtherValueTypes);
+
+        bool has_values_for_observer()
+        {
+            return std::apply([](const auto&... cached_values)
+            {
+                // Variadic expansion used
+                return (cached_values.has_value() && ...);
+            }, m_values);
+        }
+
+        auto combine_cached_values()
+        {
+            return std::apply([&](const auto&... cached_values)
+            {
+                // Variadic expansion used
```

Comment:
not sure if this comment needed =)

Reply:
Oh just a friendly reminder for someone who doesn't know about this and will become "experienced" over time.